### PR TITLE
initial port to python 3

### PR DESCRIPTION
**Pull request by jordic**

### CODE REVIEW

**Comment by jordic** [comment]
Moved from https://github.com/zopefoundation/Products.SQLAlchemyDA/pull/7

**Comment by dataflake** [comment]
Thank you for creating the branch, I'll take a look tomorrow.

**Comment by jordic** [comment]
Ideally we need to finish, merge and release z3c.sqlalchemy to be able to finish this port

File: Products/SQLAlchemyDA/da.py [hunk 1]
```diff
@@ -300,7 +300,8 @@ def getInfo(self):
         if wrapper is not None:
             d = self.sa_zope_wrapper().__dict__.copy()
             d['DSN'] = self.sa_zope_wrapper().dsn
-            for k in d.keys()[:]:
+            dd = d.copy()
+            for k in dd.keys():
                 if k.startswith('_'):
```
Comment (jugmac00) [inline Products/SQLAlchemyDA/da.py:305]: what about..
`for k in list(d.keys())`

That way you do not need to copy the complete dictionary.

**Review by icemac — CHANGES_REQUESTED** [review]
This repository still needs a `.travis.yml` file, so the tests are run.
I think it should be added in a separate PR so this one does not get messed up with these changes.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+language: python
+python:
+    - 2.7
+    - 3.8
```
Comment (dataflake) [inline .travis.yml:4]: No. Please look at other Travis configs like https://github.com/zopefoundation/Zope/blob/4.x/.travis.yml and do it that way (except the email notification)

File: Products/SQLAlchemyDA/__init__.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 
 def initialize(context):
-    from da import (SAWrapper, manage_addSAWrapper,
+    from Products.SQLAlchemyDA.da import (SAWrapper, manage_addSAWrapper,
```
Comment (dataflake) [inline Products/SQLAlchemyDA/__init__.py:19]: Make this a relative import again

**Comment by jordic** [comment]
uploaded travis, but the package depends on z3c.sqlalchemy (new version), can I pin up (like in pip) a commit on the setup.py?
There's also a failure (when running with correct version on z3c.sqlalcemy. 
https://github.com/zopefoundation/Products.SQLAlchemyDA/blob/master/tests/test_SQLAlchemyDA.py#L162

Perhaps related to, this: 
https://github.com/zopefoundation/z3c.sqlalchemy/commit/796aea500a5770f019d69f391b503cf41a1f718f
cc: @mgedmin @dataflake

File: Products/SQLAlchemyDA/da.py [hunk 4]
```diff
@@ -451,10 +451,10 @@ def manage_doQuery(self, query):
...
     def manage_formatItem(self, s):
         """ used by query form """
-        if isinstance(s, unicode):
-            return s
-        if isinstance(s, str):
-            return unicode(s, self.encoding, 'ignore')
+        # if isinstance(s, unicode):
+        #    return s
+        # if isinstance(s, str):
+        #    return unicode(s, self.encoding, 'ignore')
```
Comment (dataflake) [inline Products/SQLAlchemyDA/da.py:457]: Don't just comment this out. Look at the intent and go from there. The intent is to safely convert bytes values into unencoded string values (unicode in Python 2, native string in Python 3).

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,15 @@
+[tox]
+envlist =
+    py27,
+    py36,
```
Comment (dataflake) [inline tox.ini:4]: add py35, Zope 4 still supports it

**Review by dataflake — CHANGES_REQUESTED** [review]

**Comment by dataflake** [comment]
> uploaded travis, but the package depends on z3c.sqlalchemy (new version), can I pin up (like in pip) a commit on the setup.py?
> There's also a failure (when running with correct version on z3c.sqlalcemy.

Ignore all test failures here until the dependencies have been released.

**Comment by jordic** [comment]
When testing it localy, with an instance of zope5, I have that printed on the console
```2020-11-10 09:42:03 WARNING [SQLAlchemyDA:266][waitress-3] Unexpected failure to create SAWrapper: SAWrapper '1604183616.0983956.0.9049506331999768' already registered.
You can not register a wrapper twice under the same name.```

**Comment by dataflake** [comment]
All housekeeping stuff is now done here. The status is one failing test. That probably requires someone with actual SQLAlchemy knowledge, so it's not me.

**Comment by jordic** [comment]
Ohhh, so many thanks for the work, and the help on it!
Will try to dig on it, it's the same test that failed for me, I'm not sure, but it could be something related on the way we swapped transactions on z3c.sqlalchemy (the failing one ensures that after and insert, the session is commited). Perhaps it misses the registration? @mgedmin any ideas?

**Comment by mgedmin** [comment]
I've never used sqlalchemy in any of my Zope-related stuff, so I don't know what to tell you.

The error message hints that the z3c.sqlalchemy fix (dropping the ZopeTransactionExtension) wasn't actually correct.  It looks like it tries to call register() multiple times for the same thing.

**Comment by jordic** [comment]
Found a solution for the failure, but the problem seems to be related around the event system of sqlalchemy session.
 There are two parts, first of all:
1. session_maker it's a factory for creating sessions (the past commit on z3c.sqlalchemy) is wrong, we register the factory, not the session itself.

The second part that I'm still tring to figure why is that it needs to be registered as 'changed', if not, the session is never commited (just flushed), and that's why the test fails.

**Comment by jordic** [comment]
> The first part is that session_maker it's a factory for creating sessions (the past commit on z3c.sqlalchemy) is wrong, we register the factory, not the session itself.
> 
> https://github.com/zopefoundation/z3c.sqlalchemy/compare/fix-register-session-events

There's no need for this. The sessionmaker is also a valid param to the event system. (z3c.sqlalchemy) it's fine how it's at master.

**Comment by icemac** [comment]
I do not feel competent to review this PR.

**Comment by jordic** [comment]
@dataflake I'm currently using this package on prod for python2.7 and planning to use it with python3. Do you agree to merge this branch and release a new version.
I can keep maintaining it.

**Comment by dataflake** [comment]
This package **cannot** be released until `z3c.sqlalchemy` has been released.

**Review by dataflake — APPROVED** [review]

**Comment by jordic** [comment]
So many thanks for the help with this port :)

**Comment by jugmac00** [comment]
Thanks for contributing, @jordic!

**Comment by dataflake** [comment]
Products.SQLAlchemyDA 1.0 is now on PyPI.

**Pull request merged**